In [283]:
import pandas as pd

In [284]:
# reading the games dataset
df_games = pd.read_csv("../data/raw/games.csv")
print('Data read successfully!')

Data read successfully!


In [285]:
print(f'Dataframe shape: {df_games.shape}\n')
df_games.info()

Dataframe shape: (16715, 11)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16713 non-null  object 
 1   Platform         16715 non-null  object 
 2   Year_of_Release  16446 non-null  float64
 3   Genre            16713 non-null  object 
 4   NA_sales         16715 non-null  float64
 5   EU_sales         16715 non-null  float64
 6   JP_sales         16715 non-null  float64
 7   Other_sales      16715 non-null  float64
 8   Critic_Score     8137 non-null   float64
 9   User_Score       10014 non-null  object 
 10  Rating           9949 non-null   object 
dtypes: float64(6), object(5)
memory usage: 1.4+ MB


### dataframe first analysis
- dataframe have 16,715 records/lines;
- columns name should be updated to lowercase to avoid errors;
- NaN values should be analyzed for columns year_of_release, critic_score, user_score and rating;
- year_of_release type must be changed to int after handle NaN values;

In [286]:
# check if columns name have spaces and delete them
print(f'Column names before update:\n{df_games.columns}\n')

Column names before update:
Index(['Name', 'Platform', 'Year_of_Release', 'Genre', 'NA_sales', 'EU_sales',
       'JP_sales', 'Other_sales', 'Critic_Score', 'User_Score', 'Rating'],
      dtype='object')



In [287]:
# updating column names to snake_case
df_games.columns = df_games.columns.str.lower()
print('Column names updated to snake_case.')

Column names updated to snake_case.


In [288]:
# columns names after update
print(f'Column names after update:\n{df_games.columns}\n')

Column names after update:
Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')



In [289]:
# dataframe first 5 rows
df_games.head()

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [290]:
# checking for full duplicates
duplicates = df_games.duplicated().sum()
print(f'Number of duplicate rows: {duplicates}')

Number of duplicate rows: 0


In [291]:
# checking for missing values
missing_values = df_games.isnull().sum()
print(f'Missing values in each column:\n{missing_values}')

Missing values in each column:
name                  2
platform              0
year_of_release     269
genre                 2
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score       8578
user_score         6701
rating             6766
dtype: int64


In [292]:
# checking if missing values in 'name' and 'genre' columns are the same
print(df_games[df_games['name'].isnull()])

      name platform  year_of_release genre  na_sales  eu_sales  jp_sales  \
659    NaN      GEN           1993.0   NaN      1.78      0.53      0.00   
14244  NaN      GEN           1993.0   NaN      0.00      0.00      0.03   

       other_sales  critic_score user_score rating  
659           0.08           NaN        NaN    NaN  
14244         0.00           NaN        NaN    NaN  


### missing values
- 'name' and 'genre' column -> drop the lines as they represent really small part of df;
- 'year_of_realese' -> check if there's any patron behind this missing values; if not fill df with median;
- 'critic_score', 'user_score' and 'rating' have several missing values; will imput values using existent data.

In [293]:
# droping rows with missing values in 'name' and 'genre' columns
df_games.dropna(subset=['name', 'genre'], inplace=True)
print('Rows with missing values in "name" and "genre" columns dropped.')

Rows with missing values in "name" and "genre" columns dropped.


In [294]:
# checking sample of missing values in year_of_release column
year_null = df_games['year_of_release'].isnull()
print(df_games[year_null].sample(10))

                                                name platform  \
9706                                       Rocksmith       PC   
5294   Dragon Ball Z: Budokai Tenkaichi 2 (JP sales)      Wii   
8150                      Twisted Metal: Small Brawl       PS   
3486                                 Madden NFL 2002       XB   
12784        Tom Clancy's Rainbow Six: Critical Hour       XB   
2132                    LEGO Harry Potter: Years 5-7     X360   
5588                                Final Fantasy XI      PS2   
12209       Build-A-Bear Workshop: Friendship Valley      Wii   
5874             Harvest Moon: The Tale of Two Towns      3DS   
15953              PDC World Championship Darts 2008       DS   

       year_of_release         genre  na_sales  eu_sales  jp_sales  \
9706               NaN          Misc      0.06      0.04      0.00   
5294               NaN        Action      0.15      0.05      0.14   
8150               NaN        Action      0.10      0.07      0.00   
3486

- no patern found in missing values at 'year_of_release' column as we cannot see a systematic behaviour like "all the NaN are from specific genre or platform";
- will check if filled cell range are acurrate and fill NaN with median.

In [295]:
# checking 'year_of_release' column min and max values
print(f"Minimum year of release: {df_games['year_of_release'].min()}")
print(f"Maximum year of release: {df_games['year_of_release'].max()}")

Minimum year of release: 1980.0
Maximum year of release: 2016.0


In [296]:
# getting median year of release
median_year = df_games['year_of_release'].median()
print(f"Median year of release: {median_year}")

Median year of release: 2007.0


In [297]:
# replacing missing values in 'year_of_release' column with median value
df_games['year_of_release'].fillna(median_year, inplace=True)
print('Missing values in "year_of_release" column replaced with median value (2007).')

Missing values in "year_of_release" column replaced with median value (2007).


In [298]:
# converting 'year_of_release' column to integer type
df_games['year_of_release'] = df_games['year_of_release'].astype(int)
print(f'Column "year_of_release" type: {df_games["year_of_release"].dtype}')

Column "year_of_release" type: int32


In [299]:
# checking 'critic_score' column statistics
df_games['critic_score'].describe()

count    8137.000000
mean       68.967679
std        13.938165
min        13.000000
25%        60.000000
50%        71.000000
75%        79.000000
max        98.000000
Name: critic_score, dtype: float64

- from dataset documentation we know that critc_score max should be 100, so data is accordingly

In [300]:
# filling missing values in 'critic_score' column with mean value of the same platform and genre
df_games['critic_score'] = df_games.groupby(['platform', 'genre'])[
    'critic_score'].transform(lambda x: x.fillna(x.mean()))

In [ ]:
# repolacing NaN values in 'critic_score' column with mean value of full dataset for no samples 'platform-genre' groups
df_games['critic_score'].fillna(
    df_games['critic_score'].median(), inplace=True)

#### 'user_score' column

In [302]:
# checking 'user_score' column statistics
print(df_games['user_score'].value_counts().head(10))

user_score
tbd    2424
7.8     324
8       290
8.2     282
8.3     254
8.5     253
7.5     251
7.9     249
8.1     244
7.7     240
Name: count, dtype: int64


In [303]:
print(df_games[df_games['user_score'] == 'tbd'].sample(10))

                                 name platform  year_of_release       genre  \
6760   Classic NES Series: Excitebike      GBA             2004      Racing   
8494                  Lethal Skies II      PS2             2003  Simulation   
15880        Turbo: Super Stunt Squad      PS3             2013      Sports   
11228                   Ski and Shoot      Wii             2008      Sports   
3056               TV Show King Party      Wii             2008      Puzzle   
12981                    Domino Rally      Wii             2007      Action   
10281     Rock Band: Metal Track Pack      PS2             2009        Misc   
6247            Cabela's Trophy Bucks      PS2             2007      Sports   
12160      Major League Baseball 2K10      PSP             2010      Sports   
12065  futureU: The Prep Game for SAT       DS             2008        Misc   

       na_sales  eu_sales  jp_sales  other_sales  critic_score user_score  \
6760       0.11      0.04       0.1         0.01     

- 'tbd' represents a considerable amount of line (2,424) and no pattern was found like 'new game' or 'specific platform';
- will replace all the 'tbd' values to NaN and after will use the mean in data per platform and genre.

In [304]:
# replacing 'tbd' values in 'user_score' column with null
df_games['user_score'] = df_games['user_score'].replace('tbd', None)

In [305]:
# converting 'user_score' column to numeric type, coercing errors to null
df_games['user_score'] = pd.to_numeric(df_games['user_score'], errors='coerce')

In [306]:
# repolacing NaN values in 'user_score' column with mean value of full dataset for no samples 'platform-genre' groups
df_games['user_score'] = df_games.groupby(['platform', 'genre'])[
    'user_score'].transform(lambda x: x.fillna(x.mean()))

In [307]:
# repolacing NaN values in 'user_score' column with mean value of full dataset for no samples 'platform-genre' groups
df_games['user_score'].fillna(df_games['user_score'].median(), inplace=True)

In [308]:
print(f"Remaining NaN in user_score: {df_games['user_score'].isnull().sum()}")

Remaining NaN in user_score: 0


#### 'rating' column

In [309]:
print(df_games['rating'].value_counts())

rating
E       3990
T       2961
M       1563
E10+    1420
EC         8
K-A        3
RP         3
AO         1
Name: count, dtype: int64


In [310]:
# checking missing values in 'rating' column
missing_rating = df_games['rating'].isnull().sum()
print(f'Missing values in "rating" column: {missing_rating}')

Missing values in "rating" column: 6764


In [311]:
print(df_games[df_games['rating'].isnull()].sample(10))

                                                    name platform  \
2428                                      FIFA Soccer 08       DS   
13765                                NBA In The Zone '99      N64   
4740                      Jikkyou Powerful Pro Yakuu '94     SNES   
16376                 Dragon Quest X: All in One Package     WiiU   
2191                                      Rocksmith 2014     X360   
9932   King of Colosseum (Red): Shin Nippon x Zen Nip...      PS2   
7603                                      Snowboard Kids      N64   
15908                                     Cookie & Cream       DS   
7128                                         EX Monopoly      GBA   
13895                 Hyperdimension Neptunia Victory II      PS4   

       year_of_release         genre  na_sales  eu_sales  jp_sales  \
2428              2007        Sports      0.10      0.65      0.00   
13765             1998        Sports      0.04      0.00      0.00   
4740              1994        

- 40% of the dataset lines are with missing values, and as this is a important column for segmentation a 'Not_Rated' rating will be aplied.

In [312]:
# replacing missing values in 'rating' column with 'Not_Rated'
df_games['rating'] = df_games['rating'].fillna('Not_Rated')

In [313]:
df_games.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16713 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16713 non-null  object 
 1   platform         16713 non-null  object 
 2   year_of_release  16713 non-null  int32  
 3   genre            16713 non-null  object 
 4   na_sales         16713 non-null  float64
 5   eu_sales         16713 non-null  float64
 6   jp_sales         16713 non-null  float64
 7   other_sales      16713 non-null  float64
 8   critic_score     16713 non-null  float64
 9   user_score       16713 non-null  float64
 10  rating           16713 non-null  object 
dtypes: float64(6), int32(1), object(4)
memory usage: 1.5+ MB


In [ ]:
# adding 'total_sales' column as sum of 'na_sales', 'eu_sales', 'jp_sales', and 'other_sales' columns
df_games['total_sales'] = df_games['na_sales'] + \
    df_games['eu_sales'] + df_games['jp_sales'] + df_games['other_sales']
print('Column "total_sales" added as sum of "na_sales", "eu_sales", "jp_sales", and "other_sales" columns.')

Column "total_sales" added as sum of "na_sales", "eu_sales", "jp_sales", and "other_sales" columns.


In [315]:
# print first 5 rows of the dataframe after all transformations
print(df_games.head())

                       name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii             2006        Sports     41.36   
1         Super Mario Bros.      NES             1985      Platform     29.08   
2            Mario Kart Wii      Wii             2008        Racing     15.68   
3         Wii Sports Resort      Wii             2009        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB             1996  Role-Playing     11.27   

   eu_sales  jp_sales  other_sales  critic_score  user_score     rating  \
0     28.96      3.77         8.45     76.000000    8.000000          E   
1      3.58      6.81         0.77     68.020833    7.258442  Not_Rated   
2     12.76      3.79         3.29     82.000000    8.300000          E   
3     10.93      3.28         2.95     80.000000    8.000000          E   
4      8.89     10.22         1.00     68.020833    7.258442  Not_Rated   

   total_sales  
0        82.54  
1        40.24  
2        35